plugins/regex_revalidate/regex_revalidate.cc

```diff
@@ -0,0 +1,729 @@
+
+  if (invalidate_vec) {
+    for (Invalidate const &iv : *invalidate_vec) {
+      fprintf(stderr, "[%s] %s line: '%s' epoch: %ju expiry: %ju\n", timebuf, PLUGIN_NAME, iv.regex_text.c_str(),
+              (uintmax_t)iv.epoch, (uintmax_t)iv.expiry);
+    }
+  } else {
+    fprintf(stderr, "[%s] %s config: EMPTY\n", timebuf, PLUGIN_NAME);
+  }
+
+  fflush(stderr);
+}
+
+void
+PluginState::logConfig() const
+{
+  TSDebug(PLUGIN_NAME, "Current config: %s", config_file.c_str());
+  if (nullptr != log) {
+    TSTextLogObjectWrite(log, "Current config: %s", config_file.c_str());
+  }
+
+  if (invalidate_vec) {
+    for (Invalidate const &iv : *invalidate_vec) {
+      TSDebug(PLUGIN_NAME, "line: '%s' epoch: %ju expiry: %ju", iv.regex_text.c_str(), (uintmax_t)iv.epoch, (uintmax_t)iv.expiry);
+      if (nullptr != log) {
+        TSTextLogObjectWrite(log, "line: '%s' epoch: %ju expiry: %ju", iv.regex_text.c_str(), (uintmax_t)iv.epoch,
+                             (uintmax_t)iv.expiry);
+      }
+    }
+  } else {
+    TSDebug(PLUGIN_NAME, "Configuration EMPTY");
+    if (nullptr != log) {
+      TSTextLogObjectWrite(log, "EMPTY");
+    }
+  }
+}
+
+bool
+PluginState::fromArgs(int argc, char const **argv)
+{
+  int c;
+  constexpr option const longopts[] = {
+    {"config", required_argument, nullptr, 'c'},
+    {"disable-timed-reload", no_argument, nullptr, 'd'},
+    {"log", required_argument, nullptr, 'l'},
+    {nullptr, 0, nullptr, 0},
+  };
+
+  while ((c = getopt_long(argc, (char *const *)argv, "c:l:d", longopts, nullptr)) != -1) {
+    switch (c) {
+    case 'c':
+      config_file = optarg;
+
+      // path is relative to config dir
+      if ('/' != config_file[0]) {
+        config_file = std::string(TSConfigDirGet()) + "/" + config_file;
+      }
+
+      DEBUG_LOG("Config File: %s", config_file.c_str());
+      break;
+    case 'l':
+      if (TS_SUCCESS == TSTextLogObjectCreate(optarg, TS_LOG_MODE_ADD_TIMESTAMP, &log)) {
+        TSTextLogObjectRollingEnabledSet(log, 1);
+        TSTextLogObjectRollingIntervalSecSet(log, LOG_ROLL_INTERVAL);
+        TSTextLogObjectRollingOffsetHrSet(log, LOG_ROLL_OFFSET);
+        DEBUG_LOG("Logging Mode enabled");
+      }
+      break;
+    case 'd':
+      timed_reload = false;
+      DEBUG_LOG("Timed reload disabled (disable-timed-reload)");
+      break;
+    default:
+      break;
+    }
+  }
+
+  if (config_file.empty()) {
+    ERROR_LOG("Plugin requires a --config option along with a config file name");
+    return false;
+  }
+
+  return true;
+}
+
+time_t
+timeForFile(std::string const &filepath)
+{
+  time_t mtime{0};
+  struct stat fstat;
+  if (0 == stat(filepath.c_str(), &fstat)) {
+    mtime = fstat.st_mtime;
+  } else {
+    DEBUG_LOG("Could not stat %s", filepath.c_str());
+  }
+  return mtime;
+}
+
+// load config, true if rules changed
+bool
+PluginState::loadConfig(time_t const timenow, std::vector<Invalidate> *const rules) const
+{
+  TSAssert(nullptr != rules);
```

Comment:
Why not make rules a reference since it can't ever be null?

Reply:
c++ style.  One convention I've settled into in the past is to pass things in that may be modified by pointer and everything else in the function that is accessed by '.' methods are const.